src/subsystems/rust/translators/cargo-lock/default.nix

```diff
@@ -97,17 +97,58 @@ in {
     parsedLock = projectTree.files."Cargo.lock".tomlContent;
     parsedDeps = parsedLock.package;
 
+    makeDepId = dep: "${dep.name} ${dep.version} (${dep.source or ""})";
+
     # Gets a checksum from the [metadata] table of the lockfile
-    getChecksum = dep: let
-      key = "checksum ${dep.name} ${dep.version} (${dep.source})";
-    in
-      parsedLock.metadata."${key}";
+    getChecksum = dep: parsedLock.metadata."checksum ${makeDepId dep}";
 
     # map of dependency names to versions
     depNamesToVersions =
-      l.listToAttrs
-      (l.map (dep: l.nameValuePair dep.name dep.version) parsedDeps);
-    # This parses a "package-name version" entry in the "dependencies"
+      l.foldl'
+      (
+        all: el:
+          if l.hasAttr el.name all
+          then all // {${el.name} = all.${el.name} ++ [el];}
+          else all // {${el.name} = [el];}
+      )
+      {}
+      parsedDeps;
+    findUniqDep = let
+      # tests whether if el contains all elements from dep
+      compareAttrsets = dep: el:
```

Comment:
I think this function could be easier to understand if it was split into several ones. The nesting and inline function definitions make it a bit hard to review.

Reply:
I don't think it makes sense to split this function into more pieces, everything there is pretty much only needed for what that function does, i did add more comments though

Reply:
Thanks, the comments improve the readability a lot. Let's keep up the commenting like that. This will make the project more appealing to outside contributors.
There are some failing unit tests now. If you fix these, I think we are good to go.